### ISSUE TITLE
Vim tempfile missing after running GoCoverage

### ISSUE THREAD

**Issue by LucasRoesler** [fatih/vim-go#1663]
### What did you do? (required. The issue will be **closed** when not provided.)
Run `:GoCoverage` see some sort of error, now try `:GoFmt` and see an error like this

```
Error detected while processing function go#fmt#Format:
line   30:
E482: Can't create file /var/folders/gq/fqz3dvcs4w9d2q10dk8td7q80000gn/T/vgHcpeU/36.go
Press ENTER or type command to continue
```

All subsequent commands that require a tmp file are then broken in the same way.

I suspect that this is a Mac specific bug.  I found a workaround that fixes subsequent commands by recreating the needed tmp directory: https://stackoverflow.com/a/21100253 but it is immediately broken by running `:GoCoverage`

### What did you expect to happen?
`GoCoverage` should at least show a full error and ideally should show coverage information, `GoFmt` should succeed to fmt the file.

### What happened instead?
Multiple errors that cause all commands requiring tmp files to break.

### Configuration (**MUST** fill this out):

* Vim version (first two lines from `:version`):
```
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jan  8 2018 13:58:05)
macOS version
Included patches: 1-1400
Compiled by Homebrew
```

* Go version (`go version`):
```
go version go1.9.3 darwin/amd64
```

* Go environment (`go env`):
```
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/lucasroesler/Code/go"
GORACE=""
GOROOT="/usr/local/Cellar/go/1.9.3/libexec"
GOTOOLDIR="/usr/local/Cellar/go/1.9.3/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/gq/fqz3dvcs4w9d2q10dk8td7q80000gn/T/go-build728202914=/tmp/go-build -gno-record-gcc-switches -fno-common"
CXX="clang++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
```

* vim-go version:
current master, sha 5a205d3

* `vimrc` you used to reproduce (use a *minimal* vimrc with other plugins disabled; do not link to a 2,000 line vimrc):

```
" skip init for vim-tiny or vim-small
if 0 | endif
set encoding=utf-8


" install vim-plug if it's not already installed
if empty(glob('~/.vim/autoload/plug.vim'))
  silent !curl -fLo ~/.vim/autoload/plug.vim --create-dirs
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
  autocmd VimEnter * PlugInstall --sync | source $MYVIMRC | Startify | Ql
endif

" plugin list
call plug#begin('~/.vim/plugins')
Plug 'fatih/vim-go'
call plug#end()

" vim-go
let g:go_fmt_command = "gofmt"
let g:go_fmt_autosave = 1

" Sets the `transform` option for `gomodifytags`
let g:go_addtags_transform = "camelcase"

let g:go_highlight_types = 1
let g:go_highlight_fields = 1
let g:go_highlight_functions = 1
let g:go_highlight_methods = 1
let g:go_highlight_operators = 1
let g:go_highlight_extra_types = 1

"show the type info (|:GoInfo|) for the word under the cursor automatically.
let g:go_auto_type_info = 0
let g:go_updatetime = 800

" `gocode` should include suggestions from unimported packages
let g:go_gocode_unimported_packages = 1

let g:go_list_height = 0
let g:go_list_autoclose = 1
let g:go_statusline_duration = 6000

let g:go_metalinter_enabled = ['vet', 'golint', 'errcheck', 'interfacer']
let g:go_metalinter_autosave_enabled = ['vet', 'golint' ]
let g:go_metalinter_autosave = 0
autocmd FileType go nmap <Leader>i <Plug>(go-info)

" auto-reload this file when saving
autocmd! bufwritepost plugins.vim source %
```

**Comment by bhcleek** [comment]
What does `:messages` show after executing `:GoCoverage`?

**Comment by LucasRoesler** [comment]
```
"~/Code/go/src/github.com/lucasroesler/labs/cmd/labs/cmds/bundleDeploy.go" 96L, 2365C
vim-go: FAILED
vim-go: Add new function testfunc to the bundle
vim-go: PASS
vim-go: testing: open /var/folders/gq/fqz3dvcs4w9d2q10dk8td7q80000gn/T/v1OfFft/8: no such file or directory
vim-go: exit status 2
vim-go: FAIL        github.com/lucasroesler/labs/cmd/labs/cmds^I0.185s
vim-go: /Users/lucasroesler/Code/go/src/github.com/lucasroesler/labs
```

**Comment by Carpetsmoker** [comment]
As Christian mentioned in the Stack Overflow question you linked, this may be a Vim issue. Vim allocates a directory in the system's tmp dir per instance (or per buffer, I forgot, doesn't really matter) which will get re-used for the lifetime of that Vim process. This is an issue I've run in to on Linux as well after cleaning `/tmp` with `rmdir /tmp/*`, which will remove the (empty) directory and Vim will never re-create it. So if something if cleaning empty directories then you'll run in to problems like this.

Another posibility is that vim-go accidentally removes that directory somewhere. I've gone through the code behind `:GoCoverage` and don't see that happening anywhere though.

**Comment by LucasRoesler** [comment]
True.  I reported it here because while investigating, i found that I could consistently reproduce the bug (even within seconds of starting vim) by using the `:GoCoverage`, so I figured it must be something in vim-go or something that vim-go is using in vim that is causing the cleaning of the tmp.

**Comment by Carpetsmoker** [comment]
> I figured it must be something in vim-go or something that vim-go is using in vim that is causing the cleaning of the tmp.

Yeah, it's certainly possible, but as far as I see all it does is:

    :let tmp = tempname()
    " ... coverage code ...
    :call delete(tmp)

So I'm not sure how `:GoCoverage` would be causing this (but it's possible I missed something).

You can maybe try running the above and see if that breaks Vim in a similar way? Or try setting a different temporary directory by setting the `$TMPDIR` environment variable (see `:help tempfile`).

**Comment by LucasRoesler** [comment]
I tested it as you said and everything seems to work normally. 

I am starting to think it is a probably in my go env.  I recently ran into a odd bug while running my tests using `go test ./...`

It consistently reports 

```
# github.com/contiamo/labs/vendor/k8s.io/api/core/v1
can't create $WORK/github.com/contiamo/labs/vendor/k8s.io/api/core/v1.a: open $WORK/github.com/contiamo/labs/vendor/k8s.io/api/core/v1.a: no such file or directory
```

for various files.  If I run tests for each package individually, as reported by `go list ./...` then each test passes and it does not report any `can't create` errors.

It seems like something in the test suit is removing a directory that it shouldn't.  For `vim-go` it ends up looking like my original bug report.  If I pinpoint the exact cause and resolution, I will report it here, but at this point I think this is clearly an issue in my own test suite and we can close the issue.

**Comment by LucasRoesler** [comment]
I found my testing bug.  I will document it here just incase anyone else needs this in the future

I had many statements that look like this.
```go
tempBundleConfig,  _ := ioutil.TempFile("", "bundleTests")
bundleRoot := filepath.Dir(tempBundleConfig.Name())
defer os.RemoveAll(bundleRoot)
```

The effect is to create a new tempfile in the "standard" temp folder.  The tempfolder is then the value of `bundleRoot`. My defer was removing the standard temp folder!

I replace my `ioutil.TempFile` usage with this function to ensure that the file was in its own temp folder


```go
func tempFile(name string) (f *os.File, err error) {
	d, err := ioutil.TempDir("", "bundleTests")
	if err != nil {
		return nil, err
	}

	return ioutil.TempFile(d, name)
}
```

Once I updated all of the original temp file creation, everything is now working as expected.
